scylla-cql/src/frame/response/result.rs

```diff
@@ -373,6 +377,145 @@ impl Row {
+        &self.metadata
+    }
+
+    /// Consumes the `RawRows` and returns metadata associated with the
+    /// response.
+    #[inline]
+    pub fn into_metadata(self) -> ResultMetadata {
+        self.metadata
+    }
+
+    /// Returns the number of rows that these `RawRows` contain.
+    #[inline]
+    pub fn rows_count(&self) -> usize {
+        self.rows_count
+    }
+
+    /// Returns the serialized size of the `RawRows`.
+    #[inline]
+    pub fn rows_size(&self) -> usize {
+        self.raw_rows.len()
+    }
+
+    /// Creates a typed iterator over the rows that lazily deserializes
+    /// rows in the result.
+    ///
+    /// Returns Err if the schema of returned result doesn't match R.
+    #[inline]
+    pub fn rows_iter<'r, R: DeserializeRow<'r>>(
+        &'r self,
+    ) -> StdResult<TypedRowIterator<'r, R>, ParseError> {
+        let slice = FrameSlice::new(&self.raw_rows);
+        let raw = RowIterator::new(self.rows_count, &self.metadata.col_specs, slice);
+        TypedRowIterator::new(raw)
+    }
+
+    /// Converts the `RawRows` into `Rows` - a legacy, inefficient representation.
+    ///
+    /// Provided only to make migration to the new deserialization API
+    /// more convenient - this function will be deprecated and removed
+    /// in future releases.
+    pub fn into_legacy_rows(self) -> StdResult<Rows, ParseError> {
+        let rows = self.rows_iter::<Row>()?.collect::<StdResult<_, _>>()?;
+        Ok(Rows {
+            metadata: self.metadata,
+            rows_count: self.rows_count,
+            rows,
+            serialized_size: self.raw_rows.len(),
+        })
+    }
+}
+
+// Technically not an iterator because it returns items that borrow from it,
+// and the std Iterator interface does not allow for that.
+// TODO: Move to row.rs
+/// A _lending_ iterator over serialized rows.
+///
+/// This type is similar to `RowIterator`, but keeps ownership of the serialized
+/// result. Because it returns `ColumnIterator`s that need to borrow from it,
+/// it does not implement the `Iterator` trait (there is no type in the standard
+/// library to represent this concept yet).
+#[derive(Debug)]
+pub struct RawRowsLendingIterator {
+    metadata: ResultMetadata,
+    remaining: usize,
+    at: usize,
+    raw_rows: Bytes,
+}
+
+impl RawRowsLendingIterator {
+    /// Creates a new `RawRowsLendingIterator`, consuming given `RawRows`.
+    #[inline]
+    pub fn new(raw_rows: RawRows) -> Self {
+        Self {
+            metadata: raw_rows.metadata,
+            remaining: raw_rows.rows_count,
+            at: 0,
+            raw_rows: raw_rows.raw_rows,
+        }
+    }
+
+    /// Returns a `ColumnIterator` that represents the next row.
+    ///
+    /// Note: the `ColumnIterator` borrows from the `RawRowsLendingIterator`.
+    /// The column iterator must be consumed before the rows iterator can
+    /// continue.
+    #[inline]
+    pub fn next(&mut self) -> Option<StdResult<ColumnIterator, ParseError>> {
```

Comment:
In more detail, lifetime elision is applicable here because there is only one possible way of inferring the lifetimes: if we return a value with some lifetime ("one lifetime output"), and we provide only one parameter that has a lifetime exceeding this function's scope ("only one lifetime input"), then their lifetimes must match - that's all.